### ISSUE TITLE
How to pass weights to a custom objective function when using random forest with Scikit-Learn-Like API

### ISSUE THREAD

**Issue by flyzy2005** [dmlc/xgboost#7613]
Hello, I am trying to develop a custom objective function as below.

![image](https://user-images.githubusercontent.com/16304483/151540447-bf7a8bcf-1115-4926-b9d5-50de1a32cb0b.png)

I am using **xgboost.XGBRFRegressor**, so I created an objective function based on MSE and passed it to the _objective_ parameter.
```
def gradient_se(y_true, y_pred):
    return -2 * (y_true - y_pred)


def hessian_se(y_true, y_pred):
    return 0 * (y_true + y_pred) + 2


def custom_se(y_true, y_pred):
    grad = gradient_se(y_true, y_pred)
    hess = hessian_se(y_true, y_pred)
    return grad, hess

xgb.XGBRFRegressor(max_depth=50, objective=custom_se, random_state=1).fit(x_train, y_train)
```

As I can only get the values of `y_true `and `y_pred `in the custom objective function, I have no idea how to assign weights to grad or hess. Currently, I defined a global `WEIGHT`, and assigned it dynamicly like this, which is calculated based on `distance`.

```
global WEIGHT
dis = np.sqrt(np.sum((coords[0] - coords_train) ** 2, axis=1))
WEIGHT = np.exp(-0.5 * (dis) ** 2)
```

Then, the gradient and hessian were modified using WEIGHT:

```
def gradient_se(y_true, y_pred):
    return -2 * WEIGHT * (y_true - y_pred)


def hessian_se(y_true, y_pred):
    return 0 * (y_true + y_pred) + 2 * WEIGHT


def custom_se(y_true, y_pred):
    grad = gradient_se(y_true, y_pred)
    hess = hessian_se(y_true, y_pred)
    return grad, hess
```

Am I correct by using this method? Or is there any alternative solution to this issue?

Thanks in advance!

**Comment by trivialfis** [comment]
It's correct.

> Or is there any alternative solution to this issue?

1. Use the native interface instead, which passes a `DMatrix` object.
2. Use closure/callable instead of global variable.
``` python
# use callable
class MyObj:
    def __init__(self, weight: np.ndarray) -> None:
        self.weight = weight

    def __call__(self, y_true, y_predt) -> tuple[np.ndarray, np.ndarray]:
        return gradient(y_true, y_predt, self,weight), hessian(...)

# closure is simiar
```

**Comment by flyzy2005** [comment]
Many thanks for your suggestion! I tried callable and it worked properly.

May I ask what did you mean of 'use the native interface'?  Did you mean I can create an objective function containing a `DMatrix` object to the `objective` parameter of  `xgboost.XGBRFRegressor` like this [tutorial](https://xgboost.readthedocs.io/en/latest/tutorials/custom_metric_obj.html)?

But I read this [article](https://xgboost.readthedocs.io/en/stable/python/python_api.html) and it shows the  `objective` parameter of `xgboost.XGBRFRegressor` should be like this:

![image](https://user-images.githubusercontent.com/16304483/151552858-5a4feaed-4572-40b4-b607-6e134218fe0c.png)

**Comment by trivialfis** [comment]
Unfortunately, the RF regressor doesn't support custom objective.  Right now only gradient boosted ensemble supports it, `xgb.XGBRegressor` and `xgb.XGBClassifier`.

I saw that you have pasted a link to the "latest" doc instead of "stable".  The latest XGBoost has some improvements over the custom objective and evaluation function but hasn't been released yet.  The native interface is https://xgboost.readthedocs.io/en/latest/tutorials/custom_metric_obj.html#customized-objective-function .  From the snippet, you can see there's a `xgb.train` function.  `xgb.XGBRegressor` is a high level scikit-learn wrapper over this `train` function.

**Comment by flyzy2005** [comment]
Thanks a lot!
